File: Algorithms on Graphs/Assignment2/toposort.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

class Graph {
private:
    vector<int> visited;
    vector<vector<int>> adj;
    vector<int> topnum;
    int num;
    
    void DFS(int i) {
        visited[i] = 0;
        for (int v: adj[i]) {
            if (visited[v] == -1) {
                DFS(v);
            }
        }
        topnum[i] = num;
        num -= 1;
    }
    
public:
    Graph(vector<vector<int>>& adj_):
    adj(adj_),
    num((int)adj_.size()),
    topnum(vector<int>(adj_.size(), -1))
    {}
    
    vector<int> toposort(vector<vector<int> > adj) {
        vector<int> toposorted;
        visited = vector<int>(adj.size(), -1);
        for (int i = 0; i < adj.size(); i++) {
            if (visited[i] == -1) {
                DFS(i);
            }
        }
        vector<int> order = vector<int>(topnum.size(), -1);
        for (int i = 0; i < topnum.size(); i++) {
            order[topnum[i]-1] = i;
        }
        return order;
    }
};

int main() {
    size_t n, m;
    cin >> n >> m;
    vector<vector<int> > adj(n, vector<int>());
    for (size_t i = 0; i < m; i++) {
        int x, y;
        cin >> x >> y;
        adj[x - 1].push_back(y - 1);
    }
    Graph g = Graph(adj);
    vector<int> order = g.toposort(adj);
    for (size_t i = 0; i < order.size(); i++) {
        cout << order[i] + 1 << " ";
    }
}

File: Algorithms on Graphs/Assignment2/toposort.py
#Uses python3

import sys

class Graph:

    def __init__(self, adj):
        self.__adj = adj
        self.__visited = []
        self.__topnum = []
        self.__n = len(adj)

    def DFS(self, i):
        self.__visited[i] = 0
        for v in self.__adj[i]:
            if self.__visited[v] == -1:
                self.DFS(v)
        self.__topnum[i] = self.__n
        self.__n -= 1


    def toposort(self):
        self.__topnum = [-1 for _ in range(len(self.__adj))]
        self.__visited = [-1 for _ in range(len(self.__adj))]
        for i in range(len(adj)):
            if self.__visited[i] == -1:
                self.DFS(i)
        order = [-1 for _ in range(len(self.__adj))]
        for num in self.__topnum:
            order[num-1] = self.__topnum.index(num)
        return order


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    #data = [4, 3, 1, 2, 4, 1, 3, 1]
    #data = [5, 7, 2, 1, 3, 2, 3, 1, 4, 3, 4, 1, 5, 2, 5, 3]
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
    g = Graph(adj)
    order = g.toposort()
    for x in order:
        print(x + 1, end=' ')

